src/Stratis.Bitcoin.Features.BlockStore/BlockStore.cs

```diff
@@ -0,0 +1,363 @@
+                    await task.ConfigureAwait(false);
+                }
+                catch (OperationCanceledException)
+                {
+                    // Happens when node is shutting down or Dispose() is called.
+                    // We want to save whatever is in the batch before exiting the loop.
+                    saveBatch = true;
+
+                    this.logger.LogDebug("Node is shutting down. Save batch.");
+                }
+
+                // Save batch if timer ran out or we've dequeued a new block and reached the consensus tip 
+                // or the max batch size is reached or the node is shutting down.  
+                if (dequeueTask.Status == TaskStatus.RanToCompletion)
+                {
+                    BlockPair item = dequeueTask.Result;
+
+                    // Set the dequeue task to null so it can be assigned on the next iteration.
+                    dequeueTask = null;
+
+                    batch.Add(item);
+
+                    this.currentBatchSizeBytes += item.Block.GetSerializedSize();
+
+                    saveBatch |= (item.ChainedHeader == this.chain.Tip) || (this.currentBatchSizeBytes >= BatchThresholdSizeBytes);
+                }
+                else
+                {
+                    // Will be executed in case timer ran out or node is being shut down.
+                    saveBatch = true;
+                }
+
+                if (saveBatch)
+                {
+                    if (batch.Count != 0)
+                    {
+                        await this.SaveBatchAsync(batch).ConfigureAwait(false);
+
+                        batch.Clear();
+                        this.currentBatchSizeBytes = 0;
+                    }
+
+                    timerTask = null;
+                }
+                else
+                {
+                    // Start timer if it is not started already.
+                    timerTask = timerTask ?? Task.Delay(BatchMaxSaveIntervalSeconds * 1000, this.nodeLifetime.ApplicationStopping);
+                }
+            }
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <summary>Checks if repository contains reorged blocks and deletes them; saves batch on top. </summary>
+        /// <param name="batch">List of batched blocks. Cannot be empty.</param>
+        private async Task SaveBatchAsync(List<BlockPair> batch)
+        {
+            this.logger.LogTrace("({0}.{1}:{2})", nameof(batch), nameof(batch.Count), batch.Count);
+
+            List<BlockPair> batchCleared = this.GetBatchWithoutReorgedBlocks(batch);
+
+            ChainedHeader expectedStoreTip = batchCleared.First().ChainedHeader.Previous;
+
+            // Check if block repository contains reorged blocks. If it does - delete them.
+            if (expectedStoreTip.HashBlock != this.StoreTip.HashBlock)
+                await this.RemoveReorgedBlocksFromStoreAsync(expectedStoreTip).ConfigureAwait(false);
+
+            // Save batch.
```

Comment:
this comment doesn't match the line below

Reply:
It's a comment for 4 lines below, not the very next one

Reply:
Unusual idea to have a comment for 4 lines below, I guess you mean the comment actually covers those 4 lines of code with the final line being the action mentioned in the comment.  Perhaps a private well named method then to wrap that instead of the comment ?

Reply:
So, methodising to the same level of abstraction, as you say less converting a larger one into smaller ones for the readability I would of course agree with.

But no harm in having a single use private method to capture intent within a single block. The bonus being that the comment is redundant when you do.   However, when we look at this method here as a whole, it is doing a lot already.

so just to clarify... this is the actual save? `this.blockRepository.PutAsync`

Reply:
>this is the actual save?

Yep